### PR TITLE
show JS errors in html report

### PR DESCRIPTION
**Pull request by ngraf**
As a tester
I want to **see JavaScript errors** in **HTML reports** of failing tests
to make troubleshooting easier.

### Usage

Set `js_error_logging` to `true` for WebDriver module (opt-in).

log_js_errors_configuration

### How it looks like

js_error_in_html

### CODE REVIEW

**Comment by DavertMik** [comment]
Thanks, looks great

File: src/Codeception/Module/WebDriver.php [hunk 1]
```diff
@@ -383,11 +388,18 @@ public function debugWebDriverLogs()
...
                 );
+
                 if (empty($logEntries)) {
                     $this->debugSection("Selenium {$logType} Logs", " EMPTY ");
                     continue;
                 }
                 $this->debugSection("Selenium {$logType} Logs", "\n" . $this->formatLogEntries($logEntries));
+
+                if ($logType === 'browser' && $this->config['log_js_errors']
+                    && ($test instanceof Cest || $test instanceof Cept)
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: `$test instanceof ScenarioDriven`

File: src/Codeception/Module/WebDriver.php [hunk 2]
```diff
@@ -416,6 +428,64 @@ protected function formatLogEntries(array $logEntries)
...
 
+    /**
+     * Turns an array of log entries into a HTML string containing only relevant errors.
+     * Each log entry is an array with the keys "timestamp", "level", and "message".
+     * See https://code.google.com/p/selenium/wiki/JsonWireProtocol#Log_Entry_JSON_Object
+     *
+     * @param array $logEntries
+     * @return string
+     */
+    protected function formatLogEntriesToHtml(array $logEntries)
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: actually, why HTML? Can't this be also useful to have it inside console?
probably we shouldn't put view details and formatting inside the module?

File: src/Codeception/Module/WebDriver.php [hunk 3]
```diff
@@ -416,6 +428,64 @@ protected function formatLogEntries(array $logEntries)
...
+    {
+        if (empty($logEntries)) {
+            return '';
+        }
+
+        $formattedLogs = '
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: probably that HTML part is too specific and should be removed

File: src/Codeception/Module/WebDriver.php [hunk 4]
```diff
@@ -416,6 +428,64 @@ protected function formatLogEntries(array $logEntries)
...
+
+        $errorFound = false;
+        foreach ($logEntries as $logEntry) {
+            if ($this->isJSError($logEntry['level'], $logEntry['message'])) {
+                $errorFound = true;
+                // Timestamp is in milliseconds, but date() requires seconds.
+                $time = date('H:i:s', $logEntry['timestamp'] / 1000) .
+                    // Append the milliseconds to the end of the time string
+                    '.' . ($logEntry['timestamp'] % 1000);
+                $formattedLogs .= "{$time} {$logEntry['level']} - {$logEntry['message']}<br/>";
```
Comment (DavertMik) [inline src/Codeception/Module/WebDriver.php]: why not just:

```php
foreach ($errors as $error) {
   // time formatting happens here....
    $test->getScenario()->comment($error)
}
```

**Comment by ngraf** [comment]
Thanks for your comments @DavertMik 
I followed your advices and removed the HTML specific stuff. The feature is now more universal.

This is how it looks now in HTML:

js_errors_in_html_v2

This is how it looks now in Console:

js_errors_in_console_v2

**Comment by DavertMik** [comment]
Thanks for your PR!

**Comment by bimmelino** [comment]
thanks for code review and merging it

**Pull request merged**